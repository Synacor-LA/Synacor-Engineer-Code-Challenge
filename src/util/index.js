/**
 * Converts kelvin to fahrenheit
 * @param { number } k Temp in Kelvin
 * @returns { string } Temp in Fahrenheit
 */
export const kToFahrenheitStr = k => {
  const f = 1.8 * (k - 273.15) + 32;
  return Math.round(f).toString();
};

/** Returns an adjective to describe the temperature
 * @param { number } f The temp in Fahrenheit
 * @returns { string } A descriptor
 */
export const getDescriptor = f => {
  if (f < 32) return 'freezing';
  if (f < 60) return 'cool';
  if (f < 80) return 'balmy';
  return 'sweltering';
};
