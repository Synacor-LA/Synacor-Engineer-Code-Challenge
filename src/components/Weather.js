import React from 'react';
import { kToFahrenheitStr, getDescriptor } from '../util';

const Weather = ({ weatherData }) => {
  let { name: city } = weatherData;
  // occasionally, api will return data but not a specific city name
  if (city === '') city = 'Your detected location';
  const {
    main: { humidity, temp: kelvin },
  } = weatherData;
  const { icon, main: condition, description } = weatherData.weather[0];
  const fahrenheit = kToFahrenheitStr(kelvin);
  const descriptor = getDescriptor(fahrenheit);
  return (
    <React.Fragment>
      <div id="city">{`${city}`}</div>
      <div id="temperature">{`${fahrenheit}\u00b0 F`}</div>
      <img id="icon" src={`http://openweathermap.org/img/w/${icon}.png`} alt={description} />
      <div id="main-conditions">{`${condition}`}</div>
      <p id="blurb">
        <strong>{` ${city} `}</strong>
        {'is currently experiencing'}
        <strong>{` ${description}`}</strong>
        {`. It is a ${descriptor} ${fahrenheit} degrees, with a humidity of ${humidity}.`}
      </p>
    </React.Fragment>
  );
};

export default Weather;
