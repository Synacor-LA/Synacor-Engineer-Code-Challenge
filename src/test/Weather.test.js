import React from 'react';
import { mount, shallow } from 'enzyme';
import { createSerializer } from 'enzyme-to-json';

import Weather from '../components/Weather';
import { kToFahrenheitStr, getDescriptor } from '../util';
import weatherData from './dummy';

expect.addSnapshotSerializer(createSerializer({ mode: 'deep' }));

describe('Weather tests', () => {
  it('renders', () => {
    expect.assertions(1);
    const wrapper = shallow(<Weather weatherData={weatherData} />);
    expect(wrapper.exists()).toBe(true);
  });

  it('matches snapshot based on dummy data', () => {
    expect.assertions(1);
    const wrapper = mount(<Weather weatherData={weatherData} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('displays all text required from given weatherData object (string scan)', () => {
    expect.hasAssertions();
    // gather all relevant text from the dummy weatherData object to test for
    const fahrenheit = kToFahrenheitStr(weatherData.main.temp);
    const requiredText = [
      weatherData.name,
      weatherData.weather[0].main,
      weatherData.weather[0].description,
      weatherData.main.humidity,
      fahrenheit,
      getDescriptor(fahrenheit),
    ];
    const wrapper = mount(<Weather weatherData={weatherData} />);
    const allText = wrapper.text();
    requiredText.forEach(text => {
      expect(allText.includes(text)).toBe(true);
    });
  });
});
