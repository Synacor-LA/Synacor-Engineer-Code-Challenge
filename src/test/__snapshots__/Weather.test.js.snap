// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather tests matches snapshot based on dummy data 1`] = `
Array [
  <div
    id="city"
  >
    Marina del Rey
  </div>,
  <div
    id="temperature"
  >
    55Â° F
  </div>,
  <img
    alt="light rain"
    id="icon"
    src="http://openweathermap.org/img/w/10n.png"
  />,
  <div
    id="main-conditions"
  >
    Rain
  </div>,
  <p
    id="blurb"
  >
    <strong>
       Marina del Rey 
    </strong>
    is currently experiencing
    <strong>
       light rain
    </strong>
    . It is a cool 55 degrees, with a humidity of 66.
  </p>,
]
`;
